File: ros_time_example/scripts/ros_time_example/time_stamp_example.py
#! /usr/bin/env python3
import rospy
from geometry_msgs.msg import Twist, TwistStamped


class TimeStampExample:
    def __init__(self):
        self.pub_twist_stamped = rospy.Publisher('twist_stamped', TwistStamped, queue_size=1)
        self.sub_twist = rospy.Subscriber('twist', Twist, callback=self.recv_twist)
        self.start_time = rospy.Time.now()

    def recv_twist(self, msg):
        twist_stamped_msg = TwistStamped()

        twist_stamped_msg.header.stamp = rospy.Time.now()
        twist_stamped_msg.twist = msg

        self.pub_twist_stamped.publish(twist_stamped_msg)

        time_since_start = twist_stamped_msg.header.stamp - self.start_time
        rospy.loginfo(f'Received twist message {time_since_start.to_sec()} seconds from start')


if __name__ == '__main__':
    rospy.init_node('time_stamp_example')
    node_instance = TimeStampExample()
    rospy.spin()

File: ros_time_example/src/time_stamp_example.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <geometry_msgs/TwistStamped.h>

ros::Time start_time;
ros::Publisher pub_twist_stamped;

void recvTwist(const geometry_msgs::TwistConstPtr& msg)
{
  geometry_msgs::TwistStamped twist_stamped_msg;

  twist_stamped_msg.header.stamp = ros::Time::now();
  twist_stamped_msg.twist = *msg;

  pub_twist_stamped.publish(twist_stamped_msg);

  ros::Duration time_since_start = twist_stamped_msg.header.stamp - start_time;
  ROS_INFO_STREAM("Received twist message " << time_since_start.toSec() << " seconds from start");
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "time_stamp_example");
  ros::NodeHandle node;
  
  pub_twist_stamped = node.advertise<geometry_msgs::TwistStamped>("twist_stamped", 1);
  ros::Subscriber sub_twist = node.subscribe("twist", 1, recvTwist);

  start_time = ros::Time::now();

  ros::spin();
}